### ISSUE TITLE
catch unhandled exceptions and continue processing

### ISSUE THREAD

**Issue by vb8448** [cschneid-the-elder/mapa#147]
Hi,
i'm using the JCL parser to convert a set of files containing JCL to csv format. Not all input files contains a valid jcl and i the program halt with the error below.

Is there any way to log the errors without halting?

Thanks.

```
    Parser error line 128:11 mismatched input 'EXEC' expecting JOB
    Parser rule stack: [startRule, jcl, execJCL, jobCard]
    Exception in thread "main" java.lang.NullPointerException: Cannot invoke "org.antlr.v4.runtime.tree.TerminalNode.getSymbol()" because the return value of "JCLPPParser$JobCardContext.JOB()" is null
            at PPListener.enterJobCard(PPListener.java:66)
            at JCLPPParser$JobCardContext.enterRule(JCLPPParser.java:4975)
            at org.antlr.v4.runtime.tree.ParseTreeWalker.enterRule(ParseTreeWalker.java:50)
            at org.antlr.v4.runtime.tree.ParseTreeWalker.walk(ParseTreeWalker.java:33)
            at org.antlr.v4.runtime.tree.ParseTreeWalker.walk(ParseTreeWalker.java:36)
            at org.antlr.v4.runtime.tree.ParseTreeWalker.walk(ParseTreeWalker.java:36)
            at org.antlr.v4.runtime.tree.ParseTreeWalker.walk(ParseTreeWalker.java:36)
            at Demo01.lexAndParsePP(Demo01.java:260)
            at Demo01.main(Demo01.java:83)
```

**Comment by cschneid-the-elder** [comment]
Please provide an example that triggers the error you reported.  This will help me in testing to ensure I've found a suitable solution.

**Comment by vb8448** [comment]
Hi, here an example job

```
//JOBJOBJO JOB
//STEP0001 EXEC PGM=IEFBR14
/*
//STEP0002 EXEC PGM=IEFBR14
```

The issue seems to be the third line(`/*`), but it's a valid JCL.

BTW, the point is not this specific issue but the ability to continue the processing even in case of invalid JCL.

To bypass the problem, i've written a script that run in parallel the parser for every single file(instead of using -fileList option) but it's a lot slower because of jvm statup overhead.

**Comment by cschneid-the-elder** [comment]
Understood regarding the ability to continue processing files even when invalid JCL is encountered.  I've made several modifications which hopefully allow for this, details are in the change.log file if you're interested.

Please give the new jar in the jcl-issue-147 branch a try.  I'm going to be unable to work on this again until the middle of next week, so replies will be delayed.

**Comment by vb8448** [comment]
Hi,
I tried the new branch, the specific case I posted above now is handled, but there are some regressions with padding whites paces in the job card.

Consider this JCL:

```
//JOBJOBJO JOB (AJOBJOBJOL000)     
//STEP0001 EXEC PGM=IEFBR14
```

(after the job card there are some white spaces)

The result is:

```
FILE,test.jcl,20230226145417553,e55e7c5d-0ba7-4bb6-a165-bb9507833dc8
JOB,JOBJOBJO,1,e55e7c5d-0ba7-4bb6-a165-bb9507833dc8,ffcc363d-3620-48c6-9f61-4ed6e22b4413%     
```

Without white spaces the result is:
```
FILE,test.jcl,20230226145445675,97b062da-6e71-4114-89fd-c0d0b286f6e4
JOB,JOBJOBJO,1,97b062da-6e71-4114-89fd-c0d0b286f6e4,4bde7427-d782-443e-8c98-701bcb21a532
STEP,STEP0001,1,4bde7427-d782-443e-8c98-701bcb21a532,f202bc0d-90eb-463f-bf88-77eea5c6655b,PGM,IEFBR14% 
```

**Comment by cschneid-the-elder** [comment]
The problem is caused by my adding lexer tokens for unrecognized characters.  In at least one case, which you've discovered, I'm incorrectly missing a lexer token for the whitespace, so the whitespace isn't handled properly.  I know what I must do to fix this, but I'm going to have to defer that for a few days.  Sorry about the delay.

**Comment by vb8448** [comment]
Hi, I've done some tests and I still have some weird issues. I plan to do due further testing during the weekend.

**Comment by vb8448** [comment]
Spot another regression

```
//JOBJOBJO JOB (AJOBJOBJOL000)
//STEP0001 EXEC PGM=IEFBR14
//DD000001   DD DUMMY,DCB=(DSORG=PS,LRECL=,RECFM=)
```

With the above JCL the `jcl-issue-147` branch fails with:

```
[Fri Mar 03 10:18:02 CET 2023] WARNING: Parser error org.antlr.v4.runtime.misc.ParseCancellationException: line 3:43 no viable alternative at input '//STEP0001 EXEC PGM=IEFBR14\n//DD000001   DD DUMMY,DCB=(DSORG=PS,LRECL=,RECFM'
Exception in thread "main" java.lang.IndexOutOfBoundsException: Index 0 out of bounds for length 0
        at java.base/jdk.internal.util.Preconditions.outOfBounds(Preconditions.java:100)
        at java.base/jdk.internal.util.Preconditions.outOfBoundsCheckIndex(Preconditions.java:106)
        at java.base/jdk.internal.util.Preconditions.checkIndex(Preconditions.java:302)
        at java.base/java.util.Objects.checkIndex(Objects.java:385)
        at java.base/java.util.ArrayList.get(ArrayList.java:427)
        at Demo01.main(Demo01.java:115)
```

**Comment by cschneid-the-elder** [comment]
The IndexOutOfBoundsException is now fixed, processing of JCL with syntax errors didn't throw a fatal exception in the parser but I forgot that meant there was nothing to process afterward.  Anyway, that's fixed and there's a new jar in the jcl-issue-147 branch.

I also added more error checking in the listeners invoked during parsing, if something unrecognizable is detected then the results are considered to be invalid and thrown away.  I should really take a closer look at jobcard parsing, as it's still not perfect.

Thank you for your efforts, I do appreciate it.

**Comment by vb8448** [comment]
Hi, sorry for delay, but I didn't manage to do further tests. I hope i'll be able to do them on next Thursday.

**Comment by cschneid-the-elder** [comment]
I'm going to be unable to return to this for a couple of weeks.  Apologies.

**Comment by cschneid-the-elder** [comment]
Work has been merged into master branch.
